### ISSUE TITLE
Crash under Mono/Linux after 0.7.7.x update

### ISSUE THREAD

**Issue by kicker22004** [EcoModsDevelopers/Asphalt-ModKit#23]
I understand that Linux and ECO are iffy to say the least. But your mod used to work until 7.X and i'm wondering if you could look into the issues that are happening now.

https://pastebin.com/t7kxkMcU

Thank you, If you need a tester on linux you can reach me on the official discord under linux discussion.

**Comment by ayan4m1** [comment]
Your log looks corrupted in some way.

```
[01:28:34] Loading 0Harmony...inished in 1,730.2ms
[01:28:34] Loading Asphalt-ModKit...
[01:28:34] Loading mods finished in 1,896.0ms
[01:28:34] Server Initialization finished in 2,222.6ms5ms
```

See the _.6ms5ms_ and _inished_ rather than _finished_? Seems like something went wrong with the copy and paste or there is a more serious issue going on with the game logging in Linux.

Slightly related, is it possible that you're missing `SharpYaml.dll` in the `Mods/` directory? I would expect to see a line like `Loading SharpYaml...` but it might be due to the log oddities.

**Comment by Kronnox** [comment]
@ayan4m1 I believe that the scrambled or chopped off log output is a general problem of the Eco-Console. It probably happens because of the asynchronous initialization steps. I've actually seen it plenty of times on my own machine.

**Comment by lordmampf** [comment]
Yes the output is sometimes a little bit crazy.
But I personally don't care - so I will not do tests and improvements.
But feel free to fork and pr if you find something out ;)

I still don't understand why the SharpYaml.dll is not working with mono.
With commit 5bb0a37 I added some redirection but currently I don't have the infrastructure to test with mono. I did some research with kicker22004 and as a workaround you can place the SharpYaml.dll in the root folder if you get some exceptions. 
The reason why SharpYaml.dll is not loaded is still not clear for me. But afaik this affects only mono.

**Comment by ayan4m1** [comment]
What version of Mono are you running?

**Comment by kicker22004** [comment]
Run the most current stable buid from Mono's site. And I'm willing to be your tester dev. Please pm me.

**Comment by ayan4m1** [comment]
Able to reproduce on CentOS 7, Mono 5.16.0.187, Eco 0.7.7.2.

@lordmampf There are some issues with the changes made to try and investigate/resolve this issue. Most of the code is not having the intended effect. `Harmony.PatchAll` is causing an exception in the try/catch in the static constructor of `Asphalt.cs`, which [is then](https://github.com/EcoModsDevelopers/Asphalt-ModKit/pull/30/commits/67c34f0f65021321475662ce9d369135b547df56#diff-f1e465e8e1add9437b6c4dc2e5114c36R36) re-thrown, which causes a `TypeInitializationException` because the type failed to initialize, which snowballs into the `AggregateException` mess we're seeing with the current dev codebase running on Mono. Even on Windows the `AssemblyResolve` event handler does not have any practical use (it is only ever called for `.resource` files, which it by design skips) - it is never called under Mono (at least before the root `TypeLoadException` is thrown).

To untangle this, I am using the `UnhandledException` event of `AppDomain`. [See here](https://github.com/ayan4m1/Asphalt-ModKit/commit/8a7d6953085e92aa065cefd44c3850f62e02562d#diff-f226a492838557746b429d31bff3df8eR36) for the example. This gives us the following "root cause":

```
[04:47:30] System.TypeLoadException: Could not load type of field 'Asphalt.Storeable.Yaml.YamlSerializationHelper:Serializer' (0) due to:
Could not load file or assembly 'SharpYaml, Version=1.6.4.0, Culture=neutral, PublicKeyToken=1ced8108e942bc02' or one of its dependencies.
```

This is where I am currently stuck - SharpYaml has no oddball dependencies or issues that I can see. Harmony's docs state they have tested in Mono Linux 32- and 64-bit. Those are our only external dependencies.

Running the DLL with `Harmony.PatchAll` enabled and `MONO_LOG_LEVEL=debug` gives [this output](https://hastebin.com/yigaceluyu) - it dies around line 665:

```
Mono: Assembly Ref addref LiteDB[0x7f7b3cd49270] -> mscorlib[0x21ef680]: 50
Mono: AOT: FOUND method System.TypeInitializationException:.ctor (string,System.Exception) [0x7f7b47a34b40 - 0x7f7b47a34be7 0x7f7b47e07e2e]
```
Running the DLL with `Harmony.PatchAll` commented out (to prevent the `TypeInitializationException`) gives [this trace output](https://hastebin.com/xupacamulu) - it dies around line 590:

```
Mono: Assembly Ref addref System.Data[0x2fd2970] -> System.Numerics[0x2faa3c0]: 3
Mono: AOT: FOUND method System.TypeLoadException:.ctor (string,string) [0x7f9ad0e34f30 - 0x7f9ad0e34f51 0x7f9ad1207e58]
```

Finally, trying out what @lordmampf suggested, if I place `SharpYaml.dll` in the server root directory, I get a nasty mono crash:
```
Stacktrace:
  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) System.Reflection.Assembly.GetTypes (System.Reflection.Assembly,bool) [0x00017] in <3d1438d9da1e4fe0b915f807d1ba56bd>:0
  at System.Reflection.Assembly.GetTypes () [0x00000] in <3d1438d9da1e4fe0b915f807d1ba56bd>:0
  at Eco.Shared.Utils.ReflectionUtils/<>c__DisplayClass5_0.<DerivedTypes>b__0 (System.Reflection.Assembly) [0x00000] in <1ab4ff4856af49c58b065b4d2ea566c5>:0
```

So, sorry to report, I have spent about six hours on this so far and I'm not really any closer to solving it. Also, I'd recommend backing off of the changes made to try and fix this issue - ultimately, `Asphalt.cs` should be [this simple](https://github.com/ayan4m1/Asphalt-ModKit/blob/feature/ci/Asphalt/Asphalt.cs).

**Comment by ayan4m1** [comment]
I just noticed something in the logs and got a tiny bit further - changing `SharpYaml.dll` to `0SharpYaml.dll` changes the exception thrown. It doesn't solve the problem but it seems like load order matters - interesting to note that Harmony's DLL has a `0` prefix which would come before **A**sphalt...

**Comment by kicker22004** [comment]
Are you guys running with SharpYaml.dll in the mod folder? If so put it one folder up(root dir) and see if it works.

**Comment by DragonSupremacy** [comment]
Can confirm the server runs fine when I put the SharpYaml.dll in the root folder. Running Mono 5.16

**Comment by DragonSupremacy** [comment]
I’m running the latest version available. Not near my vm environment atm to check

**Comment by lordmampf** [comment]
removed Yaml, to this should be fixed now in version 0.5.4
